common/src/main/java/org/conscrypt/OpenSSLCipherHpke.java

```diff
@@ -0,0 +1,524 @@
+        try {
+            engineInit(opmode, key, algorithmParameterSpec, random);
+        } catch (InvalidAlgorithmParameterException e) {
+            // This shouldn't reach here, as a default spec is provided
+            throw new InvalidKeyException("Invalid Algorithm Parameters", e);
+        }
+    }
+
+    @Override
+    protected void engineInit(int opmode, Key key, AlgorithmParameterSpec params,
+            SecureRandom random) throws InvalidKeyException, InvalidAlgorithmParameterException {
+        final HpkeParameterSpec spec;
+        if (params == null && opmode == Cipher.ENCRYPT_MODE) {
+            spec = HpkeParameterSpec.DEFAULT_ENCRYPTION;
+        } else {
+            final boolean paramsInstanceOfHpke = params instanceof HpkeParameterSpec;
+            if (!paramsInstanceOfHpke) {
+                throw new InvalidAlgorithmParameterException("Only HpkeParameterSpec supported");
+            }
+            spec = (HpkeParameterSpec) params;
+        }
+
+        checkMode(spec);
+        checkAndSetOperationMode(opmode, spec);
+        checkAndSetEncodedKey(key, spec);
+        checkAndSetHpkeParams(spec);
+        checkAndSetIv(random, spec.getIv());
+        exportLength = spec.getL();
+
+        if (encrypting) {
+            enc = NativeCrypto.EVP_HPKE_CTX_setup_sender_with_seed(
+                evpCtx, evpKem, evpKdf, evpAead, encodedKey, info, randomSk);
+        } else {
+            enc = spec.getEnc();
+            evpKey = new EVP_HPKE_KEY(NativeCrypto.EVP_HPKE_KEY_new());
+            NativeCrypto.EVP_HPKE_KEY_init(evpKey, evpKem, encodedKey);
+            NativeCrypto.EVP_HPKE_CTX_setup_recipient(evpCtx, evpKey, evpKdf, evpAead, enc, info);
+        }
+        initialized = true;
+    }
+
+    @Override
+    protected void engineInit(int opmode, Key key, AlgorithmParameters params, SecureRandom random)
+            throws InvalidKeyException, InvalidAlgorithmParameterException {
+        if (params == null && opmode != Cipher.ENCRYPT_MODE) {
+            throw new InvalidAlgorithmParameterException(
+                "Only encrypt mode is supported when no AlgorithmParameters are provided");
+        }
+        final AlgorithmParameterSpec spec;
+        try {
+            spec = params == null ?
+                HpkeParameterSpec.DEFAULT_ENCRYPTION :
+                params.getParameterSpec(AlgorithmParameterSpec.class);
+        } catch (InvalidParameterSpecException e) {
+            throw new InvalidAlgorithmParameterException(
+                "AlgorithmParameters contains an invalid parameter", e);
+        }
+
+        engineInit(opmode, key, spec, random);
+    }
+
+    @Override
+    protected byte[] engineUpdate(byte[] input, int inputOffset, int inputLen) {
+        checkInitialization();
+        if (exporting) {
+            final byte[] export = NativeCrypto.EVP_HPKE_CTX_export(evpCtx, input, exportLength);
```

Comment:
Exporting is not a kind of encryption. This is definitely not the right API for HPKE.